docs/content/guides/dagster/enriching-with-software-defined-assets.mdx

```diff
@@ -0,0 +1,312 @@
+
+
+@op(ins={"users": In(Nothing)})
+def build_user_recommender_model():
+    users_df = read_sql(f"select * from users", con=create_db_connection())
+    users_recommender_model = train_recommender_model(users_df)
+    pickle_to_s3(users_recommender_model, key="users_recommender_model")
+
+
+@job
+def users_recommender_job():
+    build_user_recommender_model(build_users())
+
+
+@repository
+def repo():
+    return [users_recommender_job]
+```
+
+Below is what an equivalent job looks like with software-defined assets.
+
+```python file=/guides/dagster/enriching_with_software_defined_assets/sda_nothing.py
+from mylib import create_db_connection, pickle_to_s3, train_recommender_model
+from pandas import read_sql
+
+from dagster import AssetSelection, asset, assets_from_current_module, repository
+
+
+@asset(non_argument_deps={"raw_users"})
+def users():
+    raw_users_df = read_sql(f"select * from raw_users", con=create_db_connection())
+    users_df = raw_users_df.dropna()
+    users_df.to_sql(name="users", con=create_db_connection())
+
+
+@asset(non_argument_deps={"users"})
+def user_recommender_model():
+    users_df = read_sql(f"select * from users", con=create_db_connection())
+    users_recommender_model = train_recommender_model(users_df)
+    pickle_to_s3(users_recommender_model, key="users_recommender_model")
+
+
+@repository
+def repo():
+    assets = assets_from_current_module()
+    return [assets, AssetSelection.all().to_job("users_recommender_job")]
+```
+
+### A job where not all ops produce assets
+
+Here's a job where some of the ops - `extract_products` and `get_categories` don't producing an asset of its own - they just produce transient data that downstream ops will use to produce assets:
+
+```python file=/guides/dagster/enriching_with_software_defined_assets/vanilla_graph.py
+from pandas import DataFrame
+
+from dagster import job, op, repository
+
+from .mylib import create_db_connection, fetch_products
+
+
+@op
+def extract_products() -> DataFrame:
+    return fetch_products()
+
+
+@op
+def get_categories(products: DataFrame) -> DataFrame:
+    return DataFrame({"category": products["category"].unique()})
+
+
+@op
+def write_products_table(products: DataFrame) -> None:
+    products.to_sql(name="categories", con=create_db_connection())
+
+
+@op
+def write_categories_table(categories: DataFrame) -> None:
+    categories.to_sql(name="categories", con=create_db_connection())
+
+
+@job
+def ingest_products_and_categories():
+    products = extract_products()
+    product_categories = get_categories(products)
+    return write_products_table(products), write_categories_table(product_categories)
+
+
+@repository
+def repo():
+    return [ingest_products_and_categories]
+```
+
+Here's an equivalent job with software-defined assets. Note that, because not all the ops correspond to assets, it makes use of the `@op` and `@graph` APIs and uses `from_graph` to wrap a graph in a software-defined asset.
```

Comment:
```suggestion
Here's an equivalent job with software-defined assets. Note that because some ops don't correspond to assets, this job uses `@op` and `@graph` APIs and `from_graph` to wrap a graph in a software-defined asset:
```

I wanted to improve the clarity here, so I'm suggesting some grammar changes. Most importantly replacing `it` with `the job` so the reference is clear.